File: LeetCode/2265. Count Nodes Equal to Average of Subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int averageOfSubtree(TreeNode* root) {
        //i think we can do this using a recursive apprch
        int count = 0;
        traverse(root, count);
        return count;
    }

private:
    std::pair<int, int> traverse(TreeNode* node, int& count) {
        if (!node) {
            return {0, 0};
        }
        auto left = traverse(node->left, count);
        auto right = traverse(node->right, count);
        int sum = node->val + left.first + right.first;
        int num_nodes = 1 + left.second + right.second;
        if (node->val == sum / num_nodes) {
            count++;
        }
        return {sum, num_nodes};
    }
};

File: LeetCode/2265. Count Nodes Equal to Average of Subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int count = 0;
    //lowkey i forgot trees, so had to ask deepseek
    public int averageOfSubtree(TreeNode root) {
        dfs(root);
        return count;
    }

    // returns [subtree sum, subtree node count]
    private int[] dfs(TreeNode node) {
        if (node == null) {
            return new int[]{0, 0};
        }

        int[] left = dfs(node.left);
        int[] right = dfs(node.right);

        int sum = node.val + left[0] + right[0];
        int nodes = 1 + left[1] + right[1];

        // integer division already rounds down
        if (node.val == sum / nodes) {
            count++;
        }

        return new int[]{sum, nodes};
    }
}
